protoarray/src/main/java/tech/pegasys/teku/protoarray/ProtoArraySnaphot.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.protoarray;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.HashMap;
+import java.util.List;
+import java.util.stream.Collectors;
+import tech.pegasys.teku.util.config.Constants;
+
+public class ProtoArraySnaphot {
+
+  private final UnsignedLong justifiedEpoch;
+  private final UnsignedLong finalizedEpoch;
+  private final List<ProtoNode> nodes;
+
+  private ProtoArraySnaphot(
+      UnsignedLong justifiedEpoch, UnsignedLong finalizedEpoch, List<ProtoNode> nodes) {
+    this.justifiedEpoch = justifiedEpoch;
+    this.finalizedEpoch = finalizedEpoch;
+    this.nodes = nodes;
+  }
+
+  public static ProtoArraySnaphot save(final ProtoArray protoArray) {
+    List<ProtoNode> nodes =
+        protoArray.getNodes().stream().map(ProtoNode::cloneForSaving).collect(Collectors.toList());
+    UnsignedLong justifiedEpoch = protoArray.getJustifiedEpoch();
+    UnsignedLong finalizedEpoch = protoArray.getFinalizedEpoch();
+    return new ProtoArraySnaphot(justifiedEpoch, finalizedEpoch, nodes);
+  }
+
+  public ProtoArray toProtoArray() {
```

Comment:
If we really want to enforce immutability, we should probably create a copy of the `nodes` list we're passing into the new `ProtoArray` instance.  Or else, have `ProtoArray` create the copy internally.

Reply:
We create a copy of the `nodes` list as we're creating the `ProtoArraySnapshot`. What am I missing?

Reply:
The snapshot is indirectly mutable because you can do something like:
```
protoArray = snapshot.toProtoArray();
protoArray.onBlock(...)
```
And I think those changes will then be reflected back in the snapshot.

Reply:
Oh gotcha, makes sense.

Reply:
Actually, this is fine because it doesn't matter if the snapshot is later manipulated.

Reply:
Because the server side hands it off and does not touch it again.

Reply:
Yeah it will work but it's not a very good pattern to have a class look like it's immutable but not actually be.  Sets you up for bugs in the future when how its used changes.

Reply:
I made the comment but the next commit will make it immutable regardless.

Reply:
Done.